### ISSUE TITLE
Carts go through solid materials

### ISSUE THREAD

**Issue by Dragonop** [stujones11/minetest-railnet#3]
You can make the cart pass through solid blocks making something  similar to this:
![screenshot_20160127_122137](https://cloud.githubusercontent.com/assets/8084219/12617882/54e70b12-c4f0-11e5-8d99-c18bb7e0777e.png)

Or like this:
![screenshot_20160127_122318](https://cloud.githubusercontent.com/assets/8084219/12617959/a030e9b2-c4f0-11e5-90b6-922b4dabc09a.png)
Etc.

**Comment by stujones11** [comment]
For the first case I would say that if you make ambiguous setups like that, then you can expect ambiguous results. The second case, I agree that should not be possible, though I'm not sure how it is best fixed. I seems a shame to reduce performance by constantly checking the node around the cart just to stop people building silly things.

Thank you for reporting, I will leave the issue open for now, I'd be interested to hear any other opinions/ideas.

**Comment by HybridDog** [comment]
> […] I'm not sure how it is best fixed.

you could detect the rails touching the current one and if one is one node higher and a solid node is one node above the current rail, the current one behaves like it's not a rail

**Comment by stujones11** [comment]
> you could detect the rails touching the current one and if one is one node higher and a solid node is one node above the current rail, the current one behaves like it's not a rail

Not sure that is possible without adding complex code to predict what the 'raillike' drawtype is going to do, I guess I could find that code in the engine and see how it works but again it seems like a lot of work to prevent some quirky situation.

A simple way to deal with this would be to place a transparent node above each rail that can only be removed by digging the rail, that should prevent such nonsense ;-)

**Comment by Dragonop** [comment]
Ah, more pics then, just showing a little bit more:
![screenshot_20160127_121934](https://cloud.githubusercontent.com/assets/8084219/12628627/dd556a52-c522-11e5-8b95-d67d756f4c31.png)
![screenshot_20160127_121945](https://cloud.githubusercontent.com/assets/8084219/12628633/e38dfa42-c522-11e5-96c0-970945a67c63.png)

**Comment by stujones11** [comment]
I came up with a cheeky solution using the line_of_sight api method, I use this extensively in my shooter mod and it seems to have negligible impact on performance.

Please let me know if this has fully resolved the issue for you.

**Comment by stujones11** [comment]
Oh well, maybe a bit too cheeky, of course it does not work in unloaded map-blocks, which means they will still do that but only when no one is looking :^)

**Comment by HybridDog** [comment]
> but only when no one is looking

and the place is not forceloaded

**Comment by stujones11** [comment]
> and the place is not forceloaded

Indeed, you can forceload blocks but one of my primary objectives was to avoid that necessity

**Comment by Dragonop** [comment]
Apparently it's fixed now. Thanks stu, great mod btw, and nice new textures too.

**Comment by stujones11** [comment]
I have had to temporarily disable this feature because it very occasionally seems to give a bogus response causing carts to get stuck in the ground.
